---
output: github_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

Rocking R at UMCCR
==================

[![Travis build status](https://travis-ci.org/umccr/rock.svg?branch=master)](https://travis-ci.org/umccr/rock)
[![Coverage status](https://codecov.io/gh/umccr/rock/branch/master/graph/badge.svg)](https://codecov.io/github/umccr/rock?branch=master)

`rock` is an R package that (hopefully) helps with the day to day bioinformatics
life at UMCCR (UniMelb Centre for Cancer Research).

You can do the following:

* Create circos plots using structural variant calls from
  [Manta](https://github.com/Illumina/manta),
  and copy number variant calls from
  [CNVkit](https://github.com/etal/cnvkit),
  [FACETS](https://github.com/mskcc/facets),
  [TitanCNA](https://github.com/gavinha/TitanCNA) or
  [PURPLE](https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator).
  The OmicCircos R package is used.

* Create CNV profiles in horizontal facets for multiple samples or callers (piano plots)

## Contents

## Installation

You can install the development version of `rock` from [GitHub](https://github.com/umccr/rock) with:

```r
# install.packages("devtools") # if not pre-installed
devtools::install_github("umccr/rock") # master version
devtools::install_github("umccr/rock@v1.2.3") # release v1.2.3
devtools::install_github("umccr/rock@abcd") # commit abcd
```

There is no CRAN or conda version (yet).

Then just load with:

```{r message=FALSE}
require(rock)
```


## Circos Plots

* We can generate circos plots using the functionality available in the
  [OmicCircos](https://bioconductor.org/packages/release/bioc/html/OmicCircos.html)
  Bioconductor R package.

* Start by preparing the SV and CNV calls.

```{r circos-prep, message=FALSE}
manta <- system.file("extdata", "HCC2218_manta.vcf", package = "pebbles")
cnvkit <- system.file("extdata", "HCC2218_cnvkit-call.cns", package = "pebbles")
facets <- system.file("extdata", "HCC2218_facets_cncf.tsv", package = "pebbles")
titan <- system.file("extdata", "HCC2218_titan.segs.tsv", package = "pebbles")
purple <- system.file("extdata", "HCC2218_purple.cnv.tsv", package = "pebbles")
truth <- system.file("extdata", "HCC2218_truthset_cnv_bcbio.tsv", package = "pebbles")

sv_manta <- prep_manta_vcf(manta)
cn_facets <- prep_facets_seg(facets)
cn_cnvkit <- prep_cnvkit_seg(cnvkit)
cn_titan <- prep_titan_seg(titan)
cn_purple <- prep_purple_seg(purple)
cn_truth <- prep_truth_seg(truth)
```

* Now we can generate a circos plot with Manta links and FACETS/CNVkit/TitanCNA/PURPLE segments
  (note that the `Warning` message below is due to a hack used in the OmicCircos
  code, where a matrix with numbers and characters is coerced to numeric.
  Just don't worry about it..):

* For the internal lines:
    - The _inter_chromosomal links take the chromosome colour of mate1 of each breakend pair.
    - The _intra_chromosomal lines are coloured according to the variant type:
        - Deletions: Red
        - Duplications: Green
        - Insertions: Purple
        - Inversions: Orange


### Manta with CNVkit

```{r circos-plot-manta-cnvkit, fig.width=8, fig.height=8}
plot_circos(sv = sv_manta, cnv = cn_cnvkit)
```

### Manta with PURPLE

```{r circos-plot-manta-purple, fig.width=8, fig.height=8}
plot_circos(sv = sv_manta, cnv = cn_purple)
```

## Piano Plots

* We can generate 'piano' plots to compare CNV calls from multiple
  callers or samples.
  
```{r piano-plot-cnvkit-facets-purple-titan1, fig.width=10, fig.height=7}
plot_piano(list(truth = cn_truth, cnvkit = cn_cnvkit, facets = cn_facets, purple = cn_purple, titan = cn_titan))
```
  
  
